File: src/opspilot/memory/chunker.py
"""Markdown chunker — naive ``headings_then_size`` strategy (PR-2).

Strategy in this PR:

1. If the input begins with a YAML frontmatter (``---\\n…\\n---\\n``), strip
   it. Body coordinates remain in the *original* text's char/line space so
   chunks reference the unmodified document.
2. Walk the body line by line, tracking the ATX-heading stack
   (``# .. ######``).
3. Every heading opens a new "block"; a block's content is its heading
   line plus all subsequent non-heading lines until the next heading.
4. A block is "content-bearing" iff it has at least one non-blank line
   beyond the heading.
5. Chunks = greedy concatenation of consecutive content-bearing blocks
   until the running token estimate would exceed
   ``ChunkConfig.target_size_tokens``. Heading-only blocks are skipped
   for chunking but still update the heading-path context.

Token estimation is intentionally crude (``len(text) // 3``) — good
enough for cut-point selection. PR-5 will plug in a real tokenizer that
agrees with the embedding model's token count.

Each chunk carries:

* ``content``                 redacted markdown text
* ``char_start`` / ``char_end``    offsets into the *original* document
* ``line_start`` / ``line_end``    1-based, inclusive
* ``heading_path``            tuple of ancestor headings at chunk start
* ``anchor``                  ``#slug`` form of the leading heading
* ``token_count``             estimate
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Final, Literal

# ATX heading: 1-6 ``#`` then mandatory space then text.
_HEADING_RE: Final[re.Pattern[str]] = re.compile(r"^(#{1,6})[ \t]+(.+?)\s*$")
_FRONTMATTER_FENCE: Final[str] = "---"


@dataclass(frozen=True, slots=True)
class ChunkConfig:
    strategy: Literal["headings_then_size", "fixed_size"] = "headings_then_size"
    target_size_tokens: int = 512
    max_size_tokens: int = 1024
    overlap_tokens: int = 64


@dataclass(frozen=True, slots=True)
class Chunk:
    seq: int
    content: str
    char_start: int
    char_end: int
    line_start: int
    line_end: int
    heading_path: tuple[str, ...]
    anchor: str | None
    token_count: int


# ──────────────────────────────────────────────────────────────────────────
#  Helpers
# ──────────────────────────────────────────────────────────────────────────


def _strip_frontmatter(text: str) -> tuple[str, int, int]:
    """If text begins with ``---\\n…\\n---\\n``, return (body, body_char_start, body_line_start).

    body_line_start is 1-based — the line number of the first body line in
    the original text.
    """
    if not text.startswith(_FRONTMATTER_FENCE):
        return text, 0, 1

    lines = text.splitlines(keepends=True)
    if not lines or lines[0].rstrip("\n") != _FRONTMATTER_FENCE:
        return text, 0, 1

    for i in range(1, len(lines)):
        if lines[i].rstrip("\n") == _FRONTMATTER_FENCE:
            # Lines 0..i are frontmatter; body starts at line i+1 (0-based) → i+2 (1-based)
            body_start_char = sum(len(line) for line in lines[: i + 1])
            body = "".join(lines[i + 1 :])
            body_start_line = i + 2
            return body, body_start_char, body_start_line

    # No closing fence; treat as no frontmatter.
    return text, 0, 1


def _estimate_tokens(text: str) -> int:
    """Crude estimate: ``len(text) // 3``. Good enough for cut-point selection."""
    return max(1, len(text) // 3)


_SLUG_NON_WORD = re.compile(r"[^\w\s-]", flags=re.UNICODE)
_SLUG_DASH_SPACE = re.compile(r"[-\s]+", flags=re.UNICODE)


def _slug(text: str) -> str:
    """Convert a heading title into a URL-fragment-style anchor slug."""
    s = _SLUG_NON_WORD.sub("", text).strip().lower()
    s = _SLUG_DASH_SPACE.sub("-", s)
    return s.strip("-")


# ──────────────────────────────────────────────────────────────────────────
#  Internal block model
# ──────────────────────────────────────────────────────────────────────────


@dataclass(slots=True)
class _Block:
    """One section of the body: a heading + all lines until the next heading."""

    line_start: int  # 1-based, in original text
    line_end: int  # 1-based, inclusive
    char_start: int  # 0-based, in original text
    char_end: int  # 0-based, exclusive in original text
    content: str  # the section text (heading line included)
    heading_level: int  # 0 if no heading (preamble); 1..6 otherwise
    heading_text: str  # "" if no heading
    has_content: bool  # at least one non-blank, non-heading line
    heading_path: tuple[str, ...] = field(default_factory=tuple)


def _parse_blocks(body: str, body_char_offset: int, body_line_offset: int) -> list[_Block]:
    """Split *body* into _Block instances, tracking heading-path stack."""
    blocks: list[_Block] = []
    if not body:
        return blocks

    heading_stack: list[tuple[int, str]] = []  # [(level, text)]
    cur_lines: list[str] = []
    cur_block_start_line = body_line_offset
    cur_block_start_char = body_char_offset
    cur_heading: tuple[int, str] | None = None  # current section's leading heading

    char_pos_in_body = 0  # offset within body (not including frontmatter)
    lines = body.split("\n")

    def close_block(end_char: int, end_line: int) -> None:
        nonlocal cur_lines, cur_block_start_line, cur_block_start_char, cur_heading
        if cur_heading is None and not any(line.strip() for line in cur_lines):
            cur_lines = []
            return
        content = "\n".join(cur_lines)
        # has_content = at least one non-blank, non-heading line
        body_lines = cur_lines[1:] if cur_heading is not None and cur_lines else cur_lines
        has_content = any(line.strip() for line in body_lines)
        path = tuple(t for _, t in heading_stack)
        blocks.append(
            _Block(
                line_start=cur_block_start_line,
                line_end=end_line,
                char_start=cur_block_start_char,
                char_end=end_char,
                content=content,
                heading_level=(cur_heading[0] if cur_heading else 0),
                heading_text=(cur_heading[1] if cur_heading else ""),
                has_content=has_content,
                heading_path=path,
            )
        )
        cur_lines = []

    for line_idx, line in enumerate(lines):
        line_no = body_line_offset + line_idx
        char_pos_in_orig = body_char_offset + char_pos_in_body

        m = _HEADING_RE.match(line)
        if m:
            # Close the previous block first (its char_end is BEFORE this heading line).
            if cur_lines or cur_heading is not None:
                close_block(end_char=char_pos_in_orig, end_line=line_no - 1)
            # Update heading stack: pop all levels >= this one, then push.
            level = len(m.group(1))
            text = m.group(2).strip()
            while heading_stack and heading_stack[-1][0] >= level:
                heading_stack.pop()
            heading_stack.append((level, text))
            cur_heading = (level, text)
            cur_block_start_line = line_no
            cur_block_start_char = char_pos_in_orig
            cur_lines = [line]
        else:
            cur_lines.append(line)

        # Advance char position. ``split('\n')`` consumes the newline between
        # adjacent lines; account for the +1 except after the last line.
        char_pos_in_body += len(line)
        if line_idx < len(lines) - 1:
            char_pos_in_body += 1

    # Close the final block.
    if cur_lines or cur_heading is not None:
        final_end_line = body_line_offset + len(lines) - 1
        # If the last line is empty (trailing newline) we still treat the block
        # as ending there.
        close_block(
            end_char=body_char_offset + char_pos_in_body,
            end_line=final_end_line,
        )

    return blocks


# ──────────────────────────────────────────────────────────────────────────
#  Public entrypoint
# ──────────────────────────────────────────────────────────────────────────


def chunk_markdown(text: str, *, config: ChunkConfig | None = None) -> list[Chunk]:
    """Split *text* into chunks using the ``headings_then_size`` strategy.

    See module docstring for the full algorithm; PR-2 keeps it intentionally
    simple, deferring real tokenizer-driven sizing to PR-5.
    """
    cfg = config or ChunkConfig()
    body, body_char_offset, body_line_offset = _strip_frontmatter(text)
    blocks = _parse_blocks(body, body_char_offset, body_line_offset)

    # Greedy merge content-bearing blocks until target_size_tokens.
    chunk_groups: list[list[_Block]] = []
    pending: list[_Block] = []
    pending_tokens = 0

    for b in blocks:
        if not b.has_content:
            # Heading-only block: its heading is already in subsequent blocks'
            # heading_path (we updated the stack before creating the block).
            continue
        block_tokens = _estimate_tokens(b.content)
        if block_tokens > cfg.max_size_tokens:
            # Flush pending, emit oversize block as its own chunk.
            if pending:
                chunk_groups.append(pending)
                pending = []
                pending_tokens = 0
            chunk_groups.append([b])
            continue
        if pending and (pending_tokens + block_tokens > cfg.target_size_tokens):
            chunk_groups.append(pending)
            pending = [b]
            pending_tokens = block_tokens
        else:
            pending.append(b)
            pending_tokens += block_tokens
    if pending:
        chunk_groups.append(pending)

    # Materialize Chunk objects.
    chunks: list[Chunk] = []
    for seq, group in enumerate(chunk_groups):
        first = group[0]
        last = group[-1]

        # Reconstruct content from the original-text char range so chunks
        # are byte-faithful to the source.
        content_start = first.char_start
        content_end = last.char_end
        content = text[content_start:content_end]

        # heading_path: if all blocks share the same path, use it.
        # Otherwise compute longest common prefix and join the differing
        # tails with " / " — matches the convention used in our spec
        # examples (chunk 2 of sop_vpn_zh.md).
        if all(b.heading_path == first.heading_path for b in group):
            heading_path = first.heading_path
        else:
            min_len = min(len(b.heading_path) for b in group)
            common: list[str] = []
            for i in range(min_len):
                if all(b.heading_path[i] == group[0].heading_path[i] for b in group):
                    common.append(group[0].heading_path[i])
                else:
                    break
            tails: list[str] = []
            for b in group:
                rest = b.heading_path[len(common) :]
                if rest:
                    tails.append(rest[-1])
                elif b.heading_text:
                    tails.append(b.heading_text)
            # Deduplicate while preserving order
            seen: set[str] = set()
            unique_tails: list[str] = []
            for t in tails:
                if t not in seen:
                    unique_tails.append(t)
                    seen.add(t)
            heading_path = tuple(common) + ((" / ".join(unique_tails),) if unique_tails else ())

        anchor = "#" + _slug(first.heading_text) if first.heading_text else None
        token_count = _estimate_tokens(content)
        chunks.append(
            Chunk(
                seq=seq,
                content=content,
                char_start=first.char_start,
                char_end=last.char_end,
                line_start=first.line_start,
                line_end=last.line_end,
                heading_path=heading_path,
                anchor=anchor,
                token_count=token_count,
            )
        )

    return chunks

File: crates/opspilot-core/src/chunker.rs
/// headings_then_size chunker — Rust port of ``memory.chunker`` (PR-17).
///
/// char/byte duality: Python str uses Unicode code-point indices; Rust &str
/// uses byte offsets. We track both char_pos and byte_pos simultaneously
/// in O(1) per line. char_count() counts non-continuation bytes — LLVM
/// auto-vectorises this loop to ~16 bytes/cycle on ARM NEON / SSE2.
use std::collections::HashSet;

// ── Public types ──────────────────────────────────────────────────────────

#[derive(Debug, Clone)]
pub struct ChunkConfig {
    pub target_size_tokens: usize,
    pub max_size_tokens: usize,
}

impl Default for ChunkConfig {
    fn default() -> Self {
        Self { target_size_tokens: 512, max_size_tokens: 1024 }
    }
}

#[derive(Debug, Clone)]
pub struct Chunk {
    pub seq: usize,
    pub content: String,
    /// Unicode code-point offset (matches Python str indexing).
    pub char_start: usize,
    /// Unicode code-point offset (exclusive, matches Python str indexing).
    pub char_end: usize,
    pub line_start: usize,
    pub line_end: usize,
    pub heading_path: Vec<String>,
    pub anchor: Option<String>,
    pub token_count: usize,
}

// ── Internal block ────────────────────────────────────────────────────────

struct Block {
    line_start: usize,
    line_end: usize,
    char_start: usize,
    char_end: usize,
    byte_start: usize,
    byte_end: usize,
    heading_text: String,
    has_content: bool,
    heading_path: Vec<String>,
}

// ── Helpers ───────────────────────────────────────────────────────────────

/// Count Unicode code points in a UTF-8 string.
/// Counts non-continuation bytes (not 0b10xxxxxx). LLVM auto-vectorises
/// this into SIMD on ARM NEON and x86 SSE2 — ~16 bytes/cycle.
#[inline]
fn char_count(s: &str) -> usize {
    s.as_bytes().iter().filter(|&&b| b & 0xC0 != 0x80).count()
}

/// URL-fragment slug. Mirrors Python's _slug().
fn slug(text: &str) -> String {
    let filtered: String = text
        .chars()
        .filter(|c| c.is_alphanumeric() || *c == '_' || c.is_whitespace() || *c == '-')
        .collect();
    let lower = filtered.trim().to_lowercase();
    let mut out = String::with_capacity(lower.len());
    let mut prev_sep = false;
    for ch in lower.chars() {
        if ch.is_whitespace() || ch == '-' {
            if !prev_sep { out.push('-'); prev_sep = true; }
        } else {
            out.push(ch);
            prev_sep = false;
        }
    }
    out.trim_matches('-').to_string()
}

/// Parse an ATX heading line → (level, trimmed_text) or None.
fn parse_heading(line: &str) -> Option<(usize, &str)> {
    let b = line.as_bytes();
    let mut level = 0usize;
    while level < b.len() && level < 6 && b[level] == b'#' { level += 1; }
    if level == 0 || level >= b.len() { return None; }
    if b[level] != b' ' && b[level] != b'\t' { return None; }
    let rest = line[level..].trim();
    if rest.is_empty() { None } else { Some((level, rest)) }
}

// ── Frontmatter ───────────────────────────────────────────────────────────

/// Returns (body, body_byte_offset_in_text, body_1based_line).
pub fn strip_frontmatter(text: &str) -> (&str, usize, usize) {
    const FENCE: &str = "---";
    if !text.starts_with(FENCE) { return (text, 0, 1); }
    let mut byte_pos = 0usize;
    let mut line_num = 0usize;
    loop {
        let rem = &text[byte_pos..];
        if rem.is_empty() { break; }
        let (line_bytes, advance) = match rem.find('\n') {
            Some(nl) => (&rem[..nl], nl + 1),
            None => (rem, rem.len()),
        };
        let trimmed = line_bytes.trim_end_matches('\r');
        if line_num == 0 {
            if trimmed != FENCE { return (text, 0, 1); }
        } else if trimmed == FENCE {
            let body_start = (byte_pos + advance).min(text.len());
            return (&text[body_start..], body_start, line_num + 2);
        }
        byte_pos += advance;
        line_num += 1;
    }
    (text, 0, 1)
}

// ── Block parsing ─────────────────────────────────────────────────────────

fn make_block(
    cur_lines: &[&str],
    cur_heading: &Option<(usize, String)>,
    start_line: usize,
    char_start: usize,
    byte_start: usize,
    char_end: usize,
    byte_end: usize,
    end_line: usize,
    heading_stack: &[(usize, String)],
) -> Option<Block> {
    if cur_heading.is_none() && !cur_lines.iter().any(|l| !l.trim().is_empty()) {
        return None;
    }
    let body_start = if cur_heading.is_some() { 1 } else { 0 };
    let has_content = cur_lines[body_start..].iter().any(|l| !l.trim().is_empty());
    let heading_path: Vec<String> = heading_stack.iter().map(|(_, t)| t.clone()).collect();
    let heading_text = match cur_heading {
        Some((_, t)) => t.clone(),
        None => String::new(),
    };
    Some(Block { line_start: start_line, line_end: end_line, char_start, char_end, byte_start, byte_end, heading_text, has_content, heading_path })
}

/// Builds blocks tracking both char and byte offsets inline (no allocation).
fn parse_blocks<'a>(
    body: &'a str,
    body_byte_offset: usize,
    body_char_offset: usize,
    body_line_offset: usize,
) -> Vec<Block> {
    let mut blocks: Vec<Block> = Vec::new();
    if body.is_empty() { return blocks; }

    let mut heading_stack: Vec<(usize, String)> = Vec::new();
    let mut cur_lines: Vec<&'a str> = Vec::new();
    let mut cur_block_start_line = body_line_offset;
    let mut cur_block_char_start = body_char_offset;
    let mut cur_block_byte_start = body_byte_offset;
    let mut cur_heading: Option<(usize, String)> = None;
    let mut char_pos = body_char_offset;
    let mut byte_pos = 0usize;

    let mut lines = body.split('\n').peekable();
    let mut line_idx = 0usize;

    while let Some(line) = lines.next() {
        let is_last = lines.peek().is_none();
        let line_no = body_line_offset + line_idx;
        let orig_char = char_pos;
        let orig_byte = body_byte_offset + byte_pos;

        if let Some((level, text)) = parse_heading(line) {
            if !cur_lines.is_empty() || cur_heading.is_some() {
                if let Some(b) = make_block(
                    &cur_lines, &cur_heading,
                    cur_block_start_line,
                    cur_block_char_start, cur_block_byte_start,
                    orig_char, orig_byte,
                    line_no.saturating_sub(1),
                    &heading_stack,
                ) { blocks.push(b); }
                cur_lines.clear();
            }
            while heading_stack.last().map(|(l, _)| *l >= level).unwrap_or(false) {
                heading_stack.pop();
            }
            heading_stack.push((level, text.to_string()));
            cur_heading = Some((level, text.to_string()));
            cur_block_start_line = line_no;
            cur_block_char_start = orig_char;
            cur_block_byte_start = orig_byte;
            cur_lines.push(line);
        } else {
            cur_lines.push(line);
        }

        let line_char_len = char_count(line);
        char_pos += line_char_len;
        byte_pos += line.len();
        if !is_last { char_pos += 1; byte_pos += 1; }
        line_idx += 1;
    }

    if !cur_lines.is_empty() || cur_heading.is_some() {
        let end_byte = body_byte_offset + byte_pos;
        let end_line = body_line_offset + line_idx.saturating_sub(1);
        if let Some(b) = make_block(
            &cur_lines, &cur_heading,
            cur_block_start_line,
            cur_block_char_start, cur_block_byte_start,
            char_pos, end_byte,
            end_line,
            &heading_stack,
        ) { blocks.push(b); }
    }

    blocks
}

// ── Heading path merge ────────────────────────────────────────────────────

fn compute_heading_path(group: &[Block]) -> Vec<String> {
    let first = &group[0];
    if group.iter().all(|b| b.heading_path == first.heading_path) {
        return first.heading_path.clone();
    }
    let min_len = group.iter().map(|b| b.heading_path.len()).min().unwrap_or(0);
    let mut common_len = 0usize;
    for i in 0..min_len {
        if group.iter().all(|b| b.heading_path[i] == first.heading_path[i]) {
            common_len += 1;
        } else { break; }
    }
    let mut common = first.heading_path[..common_len].to_vec();
    let mut seen: HashSet<String> = HashSet::new();
    let mut unique_tails: Vec<String> = Vec::new();
    for b in group {
        let rest = &b.heading_path[common_len..];
        let tail = if !rest.is_empty() {
            rest.last().unwrap().clone()
        } else if !b.heading_text.is_empty() {
            b.heading_text.clone()
        } else { continue; };
        if seen.insert(tail.clone()) { unique_tails.push(tail); }
    }
    if !unique_tails.is_empty() { common.push(unique_tails.join(" / ")); }
    common
}

// ── Public entry point ────────────────────────────────────────────────────

pub fn chunk_markdown(text: &str, config: &ChunkConfig) -> Vec<Chunk> {
    let (body, body_byte_offset, body_line_offset) = strip_frontmatter(text);

    // Char offset for frontmatter end — usually just a few hundred bytes.
    let body_char_offset = char_count(&text[..body_byte_offset]);

    let blocks = parse_blocks(body, body_byte_offset, body_char_offset, body_line_offset);

    let est_tok = |char_start: usize, char_end: usize| -> usize {
        ((char_end - char_start) / 3).max(1)
    };

    let mut groups: Vec<Vec<Block>> = Vec::new();
    let mut pending: Vec<Block> = Vec::new();
    let mut pending_tokens = 0usize;

    for b in blocks {
        if !b.has_content { continue; }
        let btok = est_tok(b.char_start, b.char_end);
        if btok > config.max_size_tokens {
            if !pending.is_empty() {
                groups.push(std::mem::take(&mut pending));
                pending_tokens = 0;
            }
            groups.push(vec![b]);
        } else if !pending.is_empty() && pending_tokens + btok > config.target_size_tokens {
            groups.push(std::mem::take(&mut pending));
            pending = vec![b];
            pending_tokens = btok;
        } else {
            pending.push(b);
            pending_tokens += btok;
        }
    }
    if !pending.is_empty() { groups.push(pending); }

    groups
        .into_iter()
        .enumerate()
        .map(|(seq, group)| {
            let first = &group[0];
            let last = group.last().unwrap();
            let char_start = first.char_start;
            let char_end = last.char_end;
            let content = text[first.byte_start..last.byte_end].to_string();
            let heading_path = compute_heading_path(&group);
            let anchor = if !first.heading_text.is_empty() {
                Some(format!("#{}", slug(&first.heading_text)))
            } else { None };
            Chunk {
                seq,
                content,
                char_start,
                char_end,
                line_start: first.line_start,
                line_end: last.line_end,
                heading_path,
                anchor,
                token_count: ((char_end - char_start) / 3).max(1),
            }
        })
        .collect()
}
